// snekfetch/esm.mjs
import Snekfetch from './index.js';

export default Snekfetch;

export const version = Snekfetch.version;
export const METHODS = Snekfetch.METHODS;

export const acl = Snekfetch.acl;
export const bind = Snekfetch.bind;
export const checkout = Snekfetch.checkout;
export const connect = Snekfetch.connect;
export const copy = Snekfetch.copy;
const _delete = Snekfetch.delete;
export { _delete as delete };
export const get = Snekfetch.get;
export const head = Snekfetch.head;
export const link = Snekfetch.link;
export const lock = Snekfetch.lock;
export const merge = Snekfetch.merge;
export const mkactivity = Snekfetch.mkactivity;
export const mkcalendar = Snekfetch.mkcalendar;
export const mkcol = Snekfetch.mkcol;
export const move = Snekfetch.move;
export const notify = Snekfetch.notify;
export const options = Snekfetch.options;
export const patch = Snekfetch.patch;
export const post = Snekfetch.post;
export const propfind = Snekfetch.propfind;
export const proppatch = Snekfetch.proppatch;
export const purge = Snekfetch.purge;
export const put = Snekfetch.put;
export const rebind = Snekfetch.rebind;
export const report = Snekfetch.report;
export const search = Snekfetch.search;
export const subscribe = Snekfetch.subscribe;
export const trace = Snekfetch.trace;
export const unbind = Snekfetch.unbind;
export const unlink = Snekfetch.unlink;
export const unlock = Snekfetch.unlock;
export const unsubscribe = Snekfetch.unsubscribe;
export const brew = Snekfetch.brew;
